File: 107.binary-tree-level-order-traversal-ii.java
/*
 * @lc app=leetcode id=107 lang=java
 *
 * [107] Binary Tree Level Order Traversal II
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> res = new LinkedList<>();
        if (root == null) return res;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        res.add(0, new LinkedList());
        TreeNode nextLevelFirst = null;

        while (!queue.isEmpty()) {

            TreeNode node = queue.poll();
            if (node == nextLevelFirst) {
                res.add(0, new LinkedList());
                nextLevelFirst = null;
            }
            res.get(0).add(node.val);

            if (node.left != null) {
                queue.add(node.left);
                if (nextLevelFirst == null)
                    nextLevelFirst = node.left;
            }
            if (node.right != null) {
                queue.add(node.right);
                if (nextLevelFirst == null)
                    nextLevelFirst = node.right;
            }
        }

        return res;
    }
}

File: go/107.binary-tree-level-order-traversal-ii.go
/*
 * @lc app=leetcode id=107 lang=golang
 *
 * [107] Binary Tree Level Order Traversal II
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrderBottom(root *TreeNode) [][]int {
	res := make([][]int, 0)
	if root == nil {
		return res
	}
	curNum, nextNums := 1, 0
	queue := []*TreeNode{root}
	for len(queue) > 0 {
		tmp := make([]int, curNum)
		for i := 0; i < curNum; i++ {
			cur := queue[i]
			tmp[i] = cur.Val
			if cur.Left != nil {
				queue = append(queue, cur.Left)
				nextNums++
			}
			if cur.Right != nil {
				queue = append(queue, cur.Right)
				nextNums++
			}
		}
		queue = queue[curNum:]
		res = append(res, tmp)
		curNum, nextNums = nextNums, 0
	}
	for i, j := 0, len(res)-1; i < j; i, j = i+1, j-1 {
		res[i], res[j] = res[j], res[i]
	}
	return res
}
// @lc code=end

